### PR TITLE
Serialisation and deserialisation of cookies: failing tests, plus fixes to pass tests

### PR DESCRIPTION
**Pull request by jimsmart**
Hi,

This pull request adds tests to troubleshoot issues with both `Cookies()` and `SetCookies()`, plus the appropriate fixes for those methods to pass all the tests.

I've also added `Destroy()` to `Storage`, primarily to make all this testable, but it's probably needed anyway.

Tests have a dependency on Ginkgo and Gomega:
```bash
$ go get github.com/onsi/ginkgo/ginkgo
$ go get github.com/onsi/gomega/...
```

You will need to edit `newStore()` in `redisstorage_cookies_test.go` to point to your Redis before running the tests.

Perhaps a separate issue needs filing: you recently removed `Close()`, but you are in fact using a strategy that is based on long connections, and now your connection never gets closed. (When I said I didn't personally need it, it's because I follow a different strategy, specifically to avoid long connections, that is: in each of my methods I always obtain a new connection and close it again before returning.) — this pull request does not address this issue.

 — But let's continue the more general conversation about features / refactoring, over on https://github.com/gocolly/colly/issues/103

### CODE REVIEW

**Comment by jimsmart** [comment]
Although after much thought, I'm kinda mostly convinced that the storage plugins shouldn't actually be implementing http.CookieJar directly.

It would be better to have:-
```
type CookieStore interface {
	SetCookies(host string, cookies string) error
	Cookies(host string) (string, error)
}
```
...and then to manage the cookie jar plus any serialisation/deserialisation inside Colly. 

That would make storage plugins much simpler to implement.

It would also facilitate returning of errors from the methods that deal with cookie persistence. Currently, they were ignored (although I added logging, because it's better than ignoring stuff).

— Your thoughts?

**Comment by jimsmart** [comment]
`Destroy()` might be better named `DeleteAll()` or `RemoveAll()`.

File: redisstorage.go [hunk 1]
```diff
@@ -65,47 +86,78 @@ func (s *Storage) GetCookieJar() http.CookieJar {
...
-		if c.RawExpires != "" && !c.Expires.After(now) {
+	cnew := make([]*http.Cookie, 0, len(cookies))
+	for _, c := range cookies {
+		// Drop expired cookies.
+		if c.RawExpires != "" && c.Expires.Before(now) {
 			continue
 		}
-		cookies = append(cookies, c)
+		// Drop secure cookies if not over https.
+		if c.Secure && u.Scheme != "https" {
```
Comment (jimsmart) [inline redisstorage.go:154]: Arguably, filtering of secure cookies might not belong in this pull request?

File: redisstorage.go [hunk 2]
```diff
@@ -115,3 +167,30 @@ func (s *Storage) getIDStr(ID uint64) string {
...
+}
+
+func unstringify(s string) []*http.Cookie {
+	h := http.Header{}
+	for _, c := range strings.Split(s, "\n") {
+		h.Add("Set-Cookie", c)
+	}
+	r := http.Response{Header: h}
+	return r.Cookies()
+}
```
Comment (jimsmart) [inline redisstorage.go:187]: The previous version of the above was producing malformed cookies (each cookie became multiple garbage cookies) because it was using `http.Request`.

File: redisstorage.go [hunk 3]
```diff
@@ -65,47 +86,78 @@ func (s *Storage) GetCookieJar() http.CookieJar {
...
+
+	// Merge existing cookies, new cookies have precendence.
+	cnew := make([]*http.Cookie, len(cookies))
+	copy(cnew, cookies)
+	existing := unstringify(cookieStr)
+	for _, c := range existing {
+		if !contains(cnew, c.Name) {
+			cnew = append(cnew, c)
 		}
 	}
```
Comment (jimsmart) [inline redisstorage.go:116]: The previous version of the above was giving higher precedence to existing cookies, preventing new cookie values from being set.

**Comment by asciimoo** [comment]
First of all, thanks for the great PR.

> Although after much thought, I'm kinda mostly convinced that the storage plugins shouldn't actually be implementing http.CookieJar directly.

Yes, I agree, it makes sense. Let's do the interface modification after closing this PR.

> Destroy() might be better named DeleteAll() or RemoveAll().

I like both `DeleteAll()` and `RemoveAll()` - or perhaps `Clear()`?

File: redisstorage.go (hunk 2 above)
Comment (asciimoo) [reply redisstorage.go:187]: eh, stupid bug.. thanks for the fix.

File: redisstorage.go (hunk 1 above)
Comment (asciimoo) [reply redisstorage.go:154]: The PR fixes multiple issues, so I think it's ok here.

**Comment by asciimoo** [comment]
@jimsmart as I see you are busy nowadays, so I'm merging this PR and applying the modifications of the discussed topics. Thanks for your work!

**Comment by jimsmart** [comment]
@asciimoo apologies, yes, I've been focused on other matters recently. Thanks for completing this one. I do still fully intend to follow-up on the feature request issue I posted for Colly!

**Comment by asciimoo** [comment]
@jimsmart that's good to hear, thank you!
